datafusion/expr/src/tree_node/expr.rs

```diff
@@ -24,16 +24,13 @@ use crate::expr::{
 };
 use crate::{Expr, GetFieldAccess};
 
-use datafusion_common::tree_node::{TreeNode, VisitRecursion};
+use datafusion_common::tree_node::TreeNode;
 use datafusion_common::{internal_err, DataFusionError, Result};
 
 impl TreeNode for Expr {
-    fn apply_children<F>(&self, op: &mut F) -> Result<VisitRecursion>
-    where
-        F: FnMut(&Self) -> Result<VisitRecursion>,
-    {
-        let children = match self {
-            Expr::Alias(Alias{expr,..})
+    fn children_nodes(&self) -> Vec<Self> {
```

Comment:
This PR is a nice refactor but I'm not sure it is needed at all (1.) and I'm also not sure cloning the nodes children into a `Vec` is good direction (2.). Let me share my thoughts on these 2.

1. I think in DataFusion there are only 4 real tree types: `LogicalPlan`, `Expr`, `ExecutionPlan` and `PhysicalExpr` and there 7 special tree types based on the above 4: `SortPushDown`, `PlanWithRequitements`, `ExprOrdering`, `OrderPreservationContext`, `PipelineStatePropagator`, `PlanWithCorrespondingCoalescePartitions` and `PlanWithCorrespondingSort`.
The only reason why DataFusion currently has these 7 special trees is to be able to propagate down/up some additional information during transformation. But I think these special trees are simply not needed and can be removed. If we add some transformation helper functions like `TreeNode.transform_down_with_payload()` and `TreeNode.transform_up_with_payload()` we can easily remove these 7 special types completely.
  https://github.com/apache/arrow-datafusion/issues/8663 is about this issue and I removed `SortPushDown`, `PlanWithRequitements` and `ExprOrdering` as exaples in this PR: https://github.com/apache/arrow-datafusion/pull/8664. 

2. If we just focus on the 4 base tree types we can distingsuish 2 different types: `Expr` uses `Box`es but the other 3 (`LogicalPlan`, `ExecutionPlan` and `PhysicalExpr`) use `Arc`s for the links between the nodes. I think this is very important in their behaviour. E.g. a big issue for `Expr`s when the nodes are cloned as it basicaly  means cloning the whole subtree under the node. But cloning is not an issue for the other 3 as cloning `Arc`s is cheap. So I think the problem with the current `Expr.apply_children()` and the `Expr.children_nodes()` in this PR that it doesn't move the code into the a direction where cloning is not necessary on trees made of `Box`es.
  What I would suggest long term is in https://github.com/apache/arrow-datafusion/pull/7942. Instead of the old `TreeNode.apply_children()` (or collecting the children into a `Vec`) we could use `TreeNode.visit_children()` (https://github.com/apache/arrow-datafusion/pull/7942/files#diff-6515fda3c67b9d487ab491fd21c27e32c411a8cbee24725d737290d19c10c199R31). IMO we need to implement `visit_children()` for the 4 base tree types only (see 1.).
(Unfortunately I didn't noted the visit related changes in the https://github.com/apache/arrow-datafusion/pull/7942 PR description, but it is a big PR with many different ideas and focusing on transform seemed more important.)

Reply:
As I mentioned in above comment, the first try I did is to return `Vec<&Self>` for `children_nodes`. I think it makes more sense to return reference of children nodes instead of owned types (although for other implementations it is smart pointers, but I don't want to assume that the `TreeNode` impls must link with children nodes with `Arc` etc.). 

But there is one trouble when implement `TreeNode ` for `Arc<T>`: https://github.com/apache/arrow-datafusion/pull/8672#discussion_r1437957200. I've tried some ideas locally to get rid of it, but in the end they don't work so I changed `children_nodes` to return `Vec<Self>`.

This can be seen as an incremental cleanup patch as it doesn't change current API and keep current behavior but clean up the code as its purpose. In the long term I'd like to see more fundamental change to `TreeNode` and this can be seen as a short term effort to make the code easier to look/work on.

Reply:
Does it make sense to return `Vec<Cow<Self>>` that would contain borrowed elements for `Expr` but owned for the others?

Reply:
Yea, that's what I thought so this restrains to be limited change (mostly `apply_children`).

Reply:
> Sometimes there may be cases where one needs to do multiple passes over the same tree (e.g. doing one top down and one bottom up pass consecutively). Or we may want to run different logic on the same tree.

This is also supported with `transform_with_payload()` (https://github.com/apache/arrow-datafusion/pull/8664) that does a top-down and a bottom-up transforms too and capable to propagate different payloads in the 2 directions.

> So it may be a little premature to conclude we can do away with "derived" trees and just keep the base four.

I'm happy to try refactoring the remaining 4 (`OrderPreservationContext`, `PipelineStatePropagator`, `PlanWithCorrespondingCoalescePartitions` and `PlanWithCorrespondingSort`) in scope of https://github.com/apache/arrow-datafusion/pull/8664 next week and check the above statement.
(I thought a smaller PR that shows 1-2 examples for both up and down directions would be easier to review, but probably we can do it in 1 big PR if that's prefered.)

Reply:
I investigated the `Vec<Cow<Self>>` idea (https://github.com/apache/arrow-datafusion/pull/8672#discussion_r1438387446) a bit further and it seems possible to avoid cloning at many places. This could be especially useful for `Expr`s where cloning is very expensive. I've opened a PR to this PR to here: https://github.com/viirya/arrow-datafusion/pull/132